gene/delete.go

```diff
@@ -0,0 +1,338 @@
+package gene
+
+import (
+	"errors"
+	"github.com/vertgenlab/gonomics/dna"
+)
+
+var (
+	ErrNoStopFound = errors.New("frameshift was unable to find a stop before the end of the gene")
+)
+
+//TODO EffectPrediction
+// Deletion removes bases from the Gene, predicts the effect, and updates the Gene struct to reflect the change.
+// genomeStartPos should be the base directly BEFORE the deleted bases. genomeEndPos should be the base directly AFTER the deleted bases.
+// All positions should be given as base-zero genomic coordinates.
+func Deletion(g *Gene, genomeStartPos int, genomeEndPos int) (EffectPrediction, error) {
+	var answer EffectPrediction
+	var err error
+
+	genomeIndexStartPos, genomeIndexEndPos, err := deletionPreRunChecks(g, genomeStartPos, genomeEndPos)
+	if err != nil {
+		return EffectPrediction{}, err
+	}
+
+	// Save CDS start and stop for later
+	origCdsStart := g.cdsStarts[0]
+	origCdsEnd := g.cdsEnds[len(g.cdsEnds)-1]
+
+	// Delete from Genome Sequence
+	copy(g.genomeSeq[genomeIndexStartPos+1:], g.genomeSeq[genomeIndexEndPos:])
+	g.genomeSeq = g.genomeSeq[:len(g.genomeSeq)-(genomeIndexEndPos-(genomeIndexStartPos+1))]
+
+	// Update startPos if necessary
+	if genomeIndexStartPos == -1 {
+		g.startPos += genomeIndexEndPos - (genomeIndexStartPos + 1)
+	}
+
+	// Update CDS starts and ends and calculate how many cDNA bases have been deleted
+	var cdsIndexesToDelete []int
+	var deletedCodingBases int
+	var codingDelStart, codingDelEnd int = -1, -1
+	deletionLen := genomeIndexEndPos - (genomeIndexStartPos + 1)
+	for i := 0; i < len(g.cdsStarts); i++ {
+
+		switch {
+		// if cds is before deletion
+		case genomeIndexStartPos >= g.cdsEnds[i]:
+			codingDelStart = int(g.featureArray[g.cdsEnds[i]])
+
+		// If cds exon is fully deleted
+		case genomeIndexStartPos < g.cdsStarts[i] && genomeIndexEndPos > g.cdsEnds[i]:
+			// mark cds for deletion
+			cdsIndexesToDelete = append(cdsIndexesToDelete, i)
+			deletedCodingBases += (g.cdsEnds[i] + 1) - g.cdsStarts[i]
+			if codingDelStart == -1 {
+				codingDelStart = int(g.featureArray[g.cdsStarts[i]]) - 1
+			}
+			codingDelEnd = int(g.featureArray[g.cdsEnds[i]]) + 1
+
+		// If deletion falls within a cds
+		case genomeIndexStartPos >= g.cdsStarts[i] && genomeIndexStartPos < g.cdsEnds[i] &&
+			genomeIndexEndPos > g.cdsStarts[i] && genomeIndexEndPos <= g.cdsEnds[i]:
```

Comment:
I think the <= should be < so that it falls within?

Reply:
again, cds ends is closed

Reply:
I understand now.  I think many of my comments are explained by this misunderstanding of mine.  I would vote for zero-based left-closed right-open coordinates to be used everywhere besides specific formats/structs like gtf or vcf where we didn't have control over how that file format was originally defined.  I would also suggest we try to keep those index/interval deviations as isolated to those structs/files as possible.  I think at times other coordinate/interval systems can make things a little easier to write, but my concern is that I think it adds a level of complexity to reading/reviewing/modifying the code that can make it more difficult for someone to jump right in.  The reader needs to hold in their head the coordinate/interval system being used for each variable as well as the high-level idea of what it is holding.  I think it also adds a layer of complexity to math that involves indices/intervals that are off different systems, but being compared or used to update one of them.  I think it is also nice that golang uses the left-closed right-open in a way with how they reslice slices.